cli/cmd/edges.go

```diff
@@ -0,0 +1,325 @@
+	targets, err := util.BuildResources(options.namespace, resources)
+
+	if err != nil {
+		return nil, err
+	}
+	err = validateEdgesOutputFormat(options)
+	if err != nil {
+		return nil, err
+	}
+
+	requests := make([]*pb.EdgesRequest, 0)
+	for _, target := range targets {
+		if target.Name != "" {
+			target.Name = "" // validating that no resource name is specified in the request
+		}
+
+		requestParams := util.EdgesRequestParams{
+			ResourceType: target.Type,
+			Namespace:    options.namespace,
+		}
+
+		req, err := util.BuildEdgesRequest(requestParams)
+		if err != nil {
+			return nil, err
+		}
+		requests = append(requests, req)
+	}
+	return requests, nil
+}
+
+func edgesRespToRows(resp *pb.EdgesResponse) []*pb.Edge {
+	rows := make([]*pb.Edge, 0)
+	if resp != nil {
+		for _, edgeTable := range resp.GetOk().Edges {
+			rows = append(rows, edgeTable)
+		}
+	}
+	return rows
+}
+
+func requestEdgesFromAPI(client pb.ApiClient, req *pb.EdgesRequest) (*pb.EdgesResponse, error) {
+	resp, err := client.Edges(context.Background(), req)
+	if err != nil {
+		return nil, fmt.Errorf("Edges API error: %+v", err)
+	}
+	if e := resp.GetError(); e != nil {
+		return nil, fmt.Errorf("Edges API response error: %+v", e.Error)
+	}
+	return resp, nil
+}
+
+func renderEdgeStats(rows []*pb.Edge, options *edgesOptions) string {
+	var buffer bytes.Buffer
+	w := tabwriter.NewWriter(&buffer, 0, 0, padding, ' ', tabwriter.AlignRight)
+	writeEdgesToBuffer(rows, w, options)
+	w.Flush()
+
+	return renderEdges(buffer, options)
+}
+
+type edgeRowStats struct {
+	src    string
+	dst    string
+	client string
+	server string
+	msg    string
+}
+
+func writeEdgesToBuffer(rows []*pb.Edge, w *tabwriter.Writer, options *edgesOptions) {
+	edgeTables := make(map[string]*edgeRowStats)
+	if len(rows) != 0 {
+		for _, r := range rows {
+			key := string(r.Dst.Name + r.Src.Name)
+			clientID := r.ClientId
+			serverID := r.ServerId
+			if len(clientID) > 0 {
+				parts := strings.Split(clientID, ".")
+				clientID = parts[0] + "." + parts[1]
+			}
+			if len(serverID) > 0 {
+				parts := strings.Split(serverID, ".")
+				serverID = parts[0] + "." + parts[1]
+			}
+
+			edgeTables[key] = &edgeRowStats{
+				client: clientID,
+				server: serverID,
+				msg:    r.NoIdentityMsg,
+				src:    r.Src.Name,
+				dst:    r.Dst.Name,
+			}
+		}
+	}
+	switch options.outputFormat {
+	case tableOutput:
+		if len(edgeTables) == 0 {
+			fmt.Fprintln(os.Stderr, "No edges found.")
+			os.Exit(0)
+		}
+		printSingleEdgeTable(edgeTables, w)
+	case jsonOutput:
+		printEdgesJSON(edgeTables, w)
+	}
+}
+
+// returns the length of the longest src name
+func srcWidth(stats map[string]*edgeRowStats) int {
+	maxLength := 0
+	for _, row := range stats {
+		if len(row.src) > maxLength {
+			maxLength = len(row.src)
+		}
+	}
+	return maxLength
+}
+
+func printSingleEdgeTable(edges map[string]*edgeRowStats, w *tabwriter.Writer) {
+	// template for left-aligning the src column
```

Comment:
it seems like the headers are right aligned except for the first column:
```
$ linkerd edges po -n linkerd
SRC                                                                   DST                       CLIENT                       SERVER   MSG
linkerd-controller-5dc5958dbc-s4w2d   linkerd-prometheus-7c6d8fd47b-jhz8v   linkerd-controller.linkerd   linkerd-prometheus.linkerd     -
```
I'd expect the text column headers (and text column values) to be left aligned.

Reply:
When I run `linkerd routes deploy -n emojivoto` the first column is left-aligned but all subsequent columns are right-aligned including text columns. I do see that `linkerd top` and `linkerd endpoints` left-align all text columns. Is our style generally to left-align text? If so I can go ahead and change that.

![Screen Shot 2019-05-14 at 12 53 21 PM](https://user-images.githubusercontent.com/2289389/57727818-48581b00-7647-11e9-9a9e-c098f836c824.png)

Reply:
yea, iirc the style should be to left align text columns, and right align numbers.

Reply:
Ok cool, thanks for the spot 🔍